File: kotlin/isbn-verifier/src/main/kotlin/IsbnVerifier.kt
class IsbnVerifier {

    fun isValid(number: String): Boolean {
        var sum = 0
        var count = 0

        for (ch in number) {
            if (ch == '-') continue

            count++

            val value = when {
                ch.isDigit() -> ch - '0'
                ch == 'X' && count == 10 -> 10
                else -> return false
            }

            sum += value * (11 - count)
        }

        return count == 10 && sum % 11 == 0
    }
}

File: cpp/isbn-verifier/isbn_verifier.cpp
#include "isbn_verifier.h"

namespace isbn_verifier
{

    bool is_valid(const std::string &number)
    {
        int sum{};
        int count{};

        for (auto ch : number)
        {
            if (ch == '-')
                continue;

            count++;

            int value;

            if (isdigit(ch))
            {
                value = ch - '0';
            }
            else if (ch == 'X' && count == 10)
            {
                value = 10;
            }
            else
            {
                return false;
            }

            sum += value * (11 - count);
        }

        return count == 10 && sum % 11 == 0;
    }

} // namespace isbn_verifier
